src/MudletInstanceCoordinator.cpp

```diff
@@ -100,31 +122,61 @@ void MudletInstanceCoordinator::handleReadyRead()
     // Receive package paths and add them the install queue.
     QByteArray data = socket->readAll();
     const QStringList packagePaths = QString::fromUtf8(data).split(QChar::LineFeed, Qt::SkipEmptyParts);
-    mQueuedPackagePaths << packagePaths;
+    mQueuedUris << packagePaths;
     mMutex.unlock();
 
-    installPackagesLocally();
+    openUrisLocally();
 }
 
-// Find the active host and install queued packages to it
-void MudletInstanceCoordinator::installPackagesLocally()
+// Try to open queued URIs
+// This is called when:
+// - after a profile has been opened
+// - after another instance of Mudlet has transmitted a list of URIs to this instance
+void MudletInstanceCoordinator::openUrisLocally()
 {
+    mMutex.lock();
     QTimer::singleShot(0, this, [this]() {
         mudlet* mudletApp = mudlet::self();
         Q_ASSERT(mudletApp);
-        Host* activeHost = mudletApp->getActiveHost();
-        if (activeHost) {
-            installPackagesToHost(activeHost);
-        } else {
-            mudletApp->slot_showConnectionDialog();
+
+        QStringList skippedUris;
+
+        // Process queued URIs until we have to wait for a profile to be selected or loaded
+        while (!mQueuedUris.isEmpty()) {
+            QUrl url = QUrl(mQueuedUris.last());
+            const bool isTelnet = url.scheme() == "telnet" || url.scheme() == "mudlet";
+            const bool isPackage = url.scheme() == "file";
+            if (isTelnet) {
+                // Telnet URI is found, so we need to handle it and open a profile.
+                // Progress on uri queue will resume after the profile has been opened.
+                mQueuedUris.removeLast();
+                mMutex.unlock();
+                mudletApp->handleTelnetUri(url);
+                break;
+            }
+            if (isPackage) {
+                Host* activeHost = mudletApp->getActiveHost();
+                if (!activeHost) {
+                    // Ask the user to choose a profile, since none are active.
+                    // Progress on uri queue will resume after the profile has been opened.
+                    mudletApp->slot_showConnectionDialog();
+                    mMutex.unlock();
+                    break;
+                }
+                // Install the package to the active host
+                mQueuedUris.removeLast();
+                mMutex.unlock();
+                activeHost->installPackage(url.toLocalFile(), 0);
+            }
         }
+        mMutex.unlock();
```

Comment:
Thanks for noticing this! I placed it outside the loop to avoid checking `if(i==0)` inside the loop, but it seems like that's the correct way to do it.